man/uniunmask.1
.TH UNIUNMASK 1 "2008 Jul 6"
.SH NAME
uniunmask \- XOR bits or substitute glyphs in a GNU Unifont file.
.SH SYNOPSIS
.br
.B uniunmask
[-iinput_file] [-ooutput_file]
.SH DESCRIPTION
.B uniunmask
reads a GNU Unifont .hex file.  As it reads the font file, it scans
for code points matching entries in the "masks.hex" and "substitutes.hex"
files.  "masks.hex" and "substitutes.hex" are also in GNU Unifont .hex
format, ordered by Unicode code point (the hexadecimal number to the
left of the colon on each line).
.PP
If a code point is encountered from "substitutes.hex", the input code
point in the original .hex file is ignored (as is any entry with the
same code point in "masks.hex") and the glyph from "substitutes.hex"
is written to the output file.
.PP
Otherwise, if a code point is encountered from "masks.hex", the input
glyph in the original .hex file is XORed with the bits in the "masks.hex"
entry.  The result of this XOR operation is written to the output file.
.PP
.SH OPTIONS
.TP 12
-i
Specify the input file. The default is stdin.
.TP
-o
Specify the output file. The default is stdout.
.SH FILES
.TP 15
masks.hex, substitutes.hex, *.hex GNU Unifont font files
.SH SEE ALSO
bdfimplode(1), hex2bdf(1), hex2sfd(1), hexdraw(1), hexmerge(1), johab2ucs2(1),
unibmp2hex(1), unicoverage(1), unidup(1), unihex2bmp(1), unipagecount(1)
.SH AUTHOR
.B uniunmask
was written by Paul Hardy.
.SH LICENSE
.B uniunmask
is Copyright (C) 2007 Paul Hardy, and is released under version 2 of
the GNU General Public License, or (at your option) a later version.
.SH BUGS
No known real bugs exist, except that this software does not perform
extensive error checking on its input files.  If they're not in the
format of the original GNU Unifont hex file, all bets are off.


debian/changelog
unifont (1:5.1.20080914-1.3) unstable; urgency=low

  * Non-maintainer upload.
  * debian/rules: Split and inline "install" target into "binary-arch" and
    "binary-indep" and adjust dependencies.  (Without this,
    "dpkg-buildpackage -B" ends up running the build-indep rules and dying
    on the buildds.)
  * debian/rules: Also fix so dpkg-buildpackage -j works, and only builds
    the font files once.

 -- Samuel Bronson <naesten@gmail.com>  Sat, 12 Jan 2013 13:09:44 -0500

unifont (1:5.1.20080914-1.2) unstable; urgency=low

  [ Samuel Bronson ]
  * Non-maintainer upload.
  * Actually build the font files, using build-arch and build-indep to
    avoid killing any buildds. (Closes: #697902)
  * Don't include bogus glyphs for unallocated (as of Unicode 5.1) or
    private-use characters. (Closes: #584672)

  [ Paul Wise ]
  * Place the newly compiled fonts in precompiled dir after build.

 -- Samuel Bronson <naesten@gmail.com>  Thu, 10 Jan 2013 20:15:33 -0500

unifont (1:5.1.20080914-1.1) unstable; urgency=low

  * Non-maintainer upload to achieve release goal of getting rid
    of defoma.

 -- Christian Perrier <bubulle@debian.org>  Wed, 20 Jul 2011 22:14:06 +0200

unifont (1:5.1.20080914-1) unstable; urgency=low

  * Changed unifont.ttf location in defoma-hints file. (Closes: #498874)
  * Modified src/hex2sfd to superimpose combining diacritical marks on
    top of preceding glyphs in TrueType font as per the Unicode Standard.
    (Closes: #497326)
  * Modified "cp -a" ==> "cp -p" and eliminated "install -t" in Makefiles
    for BSD Unix compatibility.
  * Renamed "combining.txt" to "combining.dat" because it is a data file.
  * Generated new font/precompiled/unifont.ttf font from the above changes.

 -- Paul Hardy <unifoundry@unifoundry.com>  Sun,  7 Sep 2008 23:09:17 -0700

unifont (1:5.1.20080820-1) unstable; urgency=low

  * Replaced Hangul Syllables block with glyphs in unifont.bdf file
    in Debian bf-utf-source package, originally from the Debian
    xfonts-baekmuk package (Closes: #495729)
    - Avoids license ambiguity of unifont's original Hanterm fonts
    - Brings two diverging Debian copies of the unifont font closer together
  * Deleted references to and all derivative copies of any Hanterm fonts
  * Changed unifont package description in debian/control to mention
    that the package now installs /usr/share/unifont/unifont.hex and
    other files and is no longer just a metapackage (Closes: #494662)

 -- Paul Hardy <unifoundry@unifoundry.com>  Wed, 20 Aug 2008 00:08:15 -0700

unifont (1:5.1.20080808-2) unstable; urgency=low

  * Install unifont.hex [and other files while I was at it] in
    /usr/share/unifont for GRUB (Closes: #494473)
    - unifont.hex provides glyphs with no combining circles, etc. and
      is the same file as font/compiled/unifont-5.1.20080808.hex
    - unifontfull.hex provides glyphs with combining circles, etc. and
      is the same file as font/compiled/unifontfull-5.1.20080808.hex.
      If editing hex glyphs, this file is the master version to edit;
      unifont.hex is generated from this full version -- see uniunmask(1)
      for details.
    - combining.txt lists all Unicode 5.1 code points that are combining
      glyphs, and hence should have zero-width.
    - README gives a brief description of the above three files
  * Corrected typos in the previous changelog entry.

 -- Paul Hardy <unifoundry@unifoundry.com>  Sat,  9 Aug 2008 20:55:27 -0700

unifont (1:5.1.20080808-1) unstable; urgency=low

  * Added needed Replaces, Conflicts lines in debian/control (Closes: #493711)
  * Added "Build-Depends-Indep: defoma" line in debian/control for access
    to dh_installdefoma to install TrueType font (Closes: #494245)
  * Changed ttf-unifont.defoma-hints file entry for "GeneralFamily"
    from "SansSerif" to "Typewriter" as a better match
  * Changed dh_lintian lines in debian/rules to use "( ... ) || /bin/true"
    so debuild would succeed on a generic Etch DVD installation that
    doesn't include dh_lintian in the debhelper package
  * Added updates from Ubuntu unifont package:
    - hex2bdf now uses additional BDF header lines for better font
      description
    - hex2bdf-split (originally Dave Starner's adaptation of Roman
      Czyborra's hex2bdf) renamed from hex2bdf.  Use is discouraged
      but it's still here for now for completeness; wrote man page
      & changed name of output font from "-gnu-fixed-*" to "-gnu-unifont-*"
    - Added Georgian glyphs contributed to Ubuntu's unifont package by
      Gia Shervashidze and Vladimer Sichinava
  * Merged man page & Makefile changes done by Anthony Fok from Debian
    diff file into .orig.tar.gz
  * Made minor changes to existing programs in ./src
    - unibmp2hex.c: forced double-wide glyphs where appropriate in
      U+FFxx range
    - uniunmask.c: increased buffer size & hard-coded "wide" variable
      to avoid accidentally invoking a half-implemented extended
      character widths
    - unihex2bmp.c: same changes as in uniunmask.c
  * Added to existing documentation
    - Updated "./README" to reflect changes to Makefiles and add
      more information
    - Added "./font/hexsrc/README" to describe how to use all files
      in that directory to create a font
    - Added full font embedding exception to GPL from Wen Quan Yi
  * Added improved glyphs for the U+FF00..U+FFFF block
  * Removed "./font/hexsrc/blanks-5.1.hex"; it was identical to
    "blanks.hex" in the same directory and the Makefiles never used it.

 -- Paul Hardy <unifoundry@unifoundry.com>  Fri,  8 Aug 2008 02:38:05 -0700

unifont (1:5.1.20080706-1) unstable; urgency=low

  [ Paul Hardy ]
  * Adopted orphaned unifont Debian package (Closes: #356594)
  * Complete coverage of Unicode 5.1 Plane 0 (first coverage of all Plane 0)
  * New version number (5.1) reflects Unicode 5.1 as its design base
  * Adds Qianqian Fang's high-quality Wen Quan Yi CJK ideograph bitmaps

src/uniunmask.c
/*
   uniunmask - program to XOR out selected bits and substitute entire glyphs
               in a unifont.hex file

   Synopsis: uniunmask [-iin_file.hex] [-oout_file.hex]

   This program requires the files "masks.hex" and "substitutes.hex" to
   reside in the current directory.  "masks.hex" is a .hex font file with
   bits set to 1 to be XORed with an input font file.  "substitutes.hex"
   is a .hex font file with replacement glyphs for corresponding code points;
   this is mainly used to replaces space characters having graphic glyphs
   with strings of zeroes.  If a code point exists in both "substitutes.hex"
   and "masks.hex", the entry in "masks.hex" will be ignored.


   Author: Paul Hardy, unifoundry <at> unifoundry.com, 6 January 2008
   
   
    Copyright (C) 2008 Paul Hardy

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAXBUF 256




int main(int argc, char *argv[]) {

   unsigned i;                /* loop variable                     */
   unsigned k0, k1;           /* temp Unicode char variables       */
   char inbuf[256];           /* input buffer                      */
   unsigned thischar;         /* the current character             */
   unsigned char thischarbyte; /* unsigned char lowest byte of Unicode char */
   unsigned char charbits[32][4];  /* bitmap for one character, 4 bytes/row */
   unsigned charint[16];           /* bitmap as one row per entry */
   unsigned char maskbits[32][4];  /* mask for charbits[][] */
   unsigned maskint[16];           /* mask for charint[] */
   char *infile="", *outfile="";  /* names of input and output files   */
   FILE *infp, *outfp;      /* file pointers of input and output files */
   FILE *maskfp;            /* pointer to "masks.hex" file             */
   FILE *subfp;             /* pointer to "substitutes.hex" file       */
   char maskbuf[MAXBUF];    /* to hold next masked character           */
   char subbuf[MAXBUF];     /* to hold next substitute character       */
   int maskeof=0;           /* to signal no more masks to read         */
   int subeof=0;            /* to signal no more substitutes to read   */
   unsigned maskchar;       /* current mask character in buffer        */
   unsigned subchar;        /* current substitute character in buffer  */
   void hex2bit();          /* convert hex string to bitmap            */

   if ((maskfp = fopen("masks.hex", "r")) == NULL) {
      fprintf(stderr,"\nError: Can't open \"masks.hex\" for input.\n\n");
      exit(1);
   }

   if ((subfp = fopen("substitutes.hex", "r")) == NULL) {
      fprintf(stderr,"\nError: Can't open \"substitutes.hex\" for input.\n\n");
      exit(1);
   }

   if (argc > 1) {
      for (i = 1; i < argc; i++) {
         if (argv[i][0] == '-') {  /* this is an option argument */
            switch (argv[i][1]) {
               case 'i':  /* name of input file */
                  infile = &argv[i][2];
                  break;
               case 'o':  /* name of output file */
                  outfile = &argv[i][2];
                  break;
               default:   /* if unrecognized option, print list and exit */
                  fprintf(stderr, "\nSyntax:\n\n");
                  fprintf(stderr, "   %s -p<Unicode_Page> ", argv[0]);
                  fprintf(stderr, "-i<Input_File> -o<Output_File> -w\n\n");
                  fprintf(stderr, "\nExample:\n\n");
                  exit(1);
            }
         }
      }
   }
   /*
      Make sure we can open any I/O files that were specified before
      doing anything else.
   */
   if (strlen(infile) > 0) {
      if ((infp = fopen(infile, "r")) == NULL) {
         fprintf(stderr, "Error: can't open %s for input.\n", infile);
         exit(1);
      }
   }
   else {
      infp = stdin;
   }
   if (strlen(outfile) > 0) {
      if ((outfp = fopen(outfile, "w")) == NULL) {
         fprintf(stderr, "Error: can't open %s for output.\n", outfile);
         exit(1);
      }
   }
   else {
      outfp = stdout;
   }
   /*
      Get first mask code point.
   */
   if (fgets(maskbuf, MAXBUF-1, maskfp) == NULL) {
      maskeof = 1;
   }
   else {
      sscanf(maskbuf, "%x", &maskchar); /* get next mask code point */
   }
   /*
      Get first substitute code point.
   */
   if (fgets(subbuf, MAXBUF-1, subfp) == NULL) {
      subeof = 1;
   }
   else {
      sscanf(subbuf, "%x", &subchar); /* get next substitute code point */
   }
   /*
      Read in the glyphs in the file
   */
   while (fgets(inbuf, MAXBUF-1, infp) != NULL) {
      sscanf(inbuf, "%x", &thischar);
      thischarbyte = (unsigned char)(thischar & 0xff);
      /*
         Advance to the right position in the masks.hex file.
         It's okay if we go to EOF.
      */
      while (!maskeof && thischar > maskchar) {
         if (fgets(maskbuf, MAXBUF-1, maskfp) == NULL) {
            maskeof = 1;
         }
         else {
            sscanf(maskbuf, "%x", &maskchar); /* get next mask code point */
         }
      }
      /*
         Advance to the right position in the substitute.hex file.
         It's okay if we go to EOF.

         Note that ALL entries in substitute.hex are printed, unlike
         all entries in masks.hex.  This guarantees that we handle all
         space characters properly for a font regardless of whether or
         not it is in the input .hex file.

         If substitute code point has no hex string, delete the input
         character from the output.
      */
      while (!subeof && thischar > subchar) {
         if (fgets(subbuf, MAXBUF-1, subfp) == NULL) {
            subeof = 1;
         }
         else {
            sscanf(subbuf, "%x", &subchar); /* get next substitute code point */
            if (subchar < thischar) {
               fprintf(outfp, "%s", subbuf); /* print every sub code point */
            }
         }
      }
      if (thischar == subchar) {
         if (strlen(subbuf) >= 37) {
            fprintf(outfp, "%s", subbuf); /* print substitute character */
         }
         /* otherwise, do nothing -- delete input code point from output */
      }
      else if (maskeof || thischar < maskchar) {
         fprintf(outfp, "%s", inbuf); /* print original character */
      }
      /*
         If the character should have a combining circle, look for it
         and generate a new hex string without it.
      */
      else {
         for (k0=0; inbuf[k0] != ':'; k0++);
         k0++;
         hex2bit(&inbuf[k0], charbits); /* convert hex string to 32*4 bitmap */
         for (k1=0; maskbuf[k1] != ':'; k1++);
         k1++;
         hex2bit(&maskbuf[k1], maskbits);
         for (i=0; i<16; i++) {
            /* combine two bytes of a glyph row into one int */
            charint[i] = (charbits[i+8][1] << 8) | charbits[i+8][2];
            /* combine two bytes of a mask row into one int */
            maskint[i] = (maskbits[i+8][1] << 8) | maskbits[i+8][2];
            charint[i] ^= maskint[i]; /* turn off bits that were 1 in mask */
         }
         /*
            Now print the glyph hex string with the combining circle removed.
         */
         fprintf(outfp, "%04X:", thischar);
         if (strlen(inbuf) > 63) { /* full-width character */
            for (i=0; i<16; i++) {
               thischarbyte = (charint[i] >> 8) & 0xff;
               fprintf(outfp, "%02X", thischarbyte);
               thischarbyte = charint[i] & 0xff;
               fprintf(outfp, "%02X", thischarbyte);
            }
            fprintf(outfp, "\n");
         }
         else { /* half-width character */
            for (i=0; i<16; i++) {
               thischarbyte = (charint[i] >> 8) & 0xff;
               fprintf(outfp, "%02X", thischarbyte);
            }
            fprintf(outfp, "\n");
         }
      }
   }
   exit(0);
}

/*
   Convert the portion of a hex string after the ':' into a character bitmap.

   If string is >= 128 characters, it will fill all 4 bytes per row.
   If string is >= 64 characters and < 128, it will fill 2 bytes per row.
   Otherwise, it will fill 1 byte per row.
*/
void hex2bit(char *instring, unsigned char character[32][4]) {

   int i;  /* current row in bitmap character */
   int j;  /* current character in input string */
   int k;  /* current byte in bitmap character */
   int width;  /* number of output bytes to fill: 1, 2, or 4 */

   for (i=0; i<32; i++)  /* erase previous character */
      character[i][0] = character[i][1] = character[i][2] = character[i][3] = 0; 
   j=0;  /* current location is at beginning of instring */

   // width = (strlen(instring) - 1) >> 4; /* 16 hex digits per 8 bytes */

   if (strlen(instring) <= 34)  /* 32 + possible '\r', '\n' */
      width = 0;
   else if (strlen(instring) <= 66)  /* 64 + possible '\r', '\n' */
      width = 1;
   else
      width = 4;

   k = (width > 1) ? 0 : 1;  /* if width < 3, start at index 1 else at 0 */

   for (i=8; i<24; i++) {  /* 16 rows per input character, rows 8..23 */
      sscanf(&instring[j], "%2hhx", &character[i][k]);
      j += 2;
      if (width > 0) { /* add next pair of hex digits to this row */
         sscanf(&instring[j], "%2hhx", &character[i][k+1]);
         j += 2;
      }
      if (width > 1) { /* add 2 next pairs of hex digits to this row */
         sscanf(&instring[j], "%2hhx", &character[i][k+2]);
         j += 2;
         sscanf(&instring[j], "%2hhx", &character[i][k+3]);
         j += 2;
      }
   }

   return;
}

/*
   init() -- read the file "combining.dat" and populate the combining[]
             array with a list of code points that contain combining circles.
*/
int init(unsigned char combining[17*65536]) {

   int i, j;
   char inbuf[MAXBUF];
   FILE *combiningfp;

   if ((combiningfp = fopen("combining.dat", "r")) == NULL) {
      fprintf(stderr, "\nError: data file \"combining.dat\" not found.\n\n");
      exit(0);
   }

   while (fgets(inbuf, MAXBUF-1, combiningfp) != NULL) {
      if ((inbuf[0] >= '0' && inbuf[0] <= '9') ||
          (inbuf[0] >= 'A' && inbuf[0] <= 'F') ||
          (inbuf[0] >= 'a' && inbuf[0] <= 'f')) {
         j = 0;  /* in case second hex number isn't there */
         sscanf(inbuf, "%x-%x", &i, &j);
         do {
            combining[i] = 1;
            i++;
         } while (i <= j);
      }
   }
   fclose(combiningfp);

   return(0);
}
